src/Avalonia.Controls/PullToRefresh/RefreshVisualizer.cs

```diff
@@ -0,0 +1,647 @@
+                    }
+                    else
+                    {
+                        RefreshVisualizerState = RefreshVisualizerState.Idle;
+                    }
+                }
+            }
+        }
+
+        private void InteractingForRefreshObserver(bool obj)
+        {
+            _isInteractingForRefresh = obj;
+
+            if (!_isInteractingForRefresh)
+            {
+                switch (_refreshVisualizerState)
+                {
+                    case RefreshVisualizerState.Pending:
+                        RequestRefresh();
+                        break;
+                    case RefreshVisualizerState.Refreshing:
+                        // We don't want to interrupt a currently executing refresh.
+                        break;
+                    default:
+                        RefreshVisualizerState = RefreshVisualizerState.Idle;
+                        break;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Defines constants that specify the state of a RefreshVisualizer
+    /// </summary>
+    public enum RefreshVisualizerState
+    {
+        Idle,
+        Peeking,
+        Interacting,
+        Pending,
+        Refreshing
+    }
+
+    /// <summary>
+    /// Defines constants that specify the orientation of a RefreshVisualizer.
+    /// </summary>
+    public enum RefreshVisualizerOrientation
+    {
+        Auto,
+        Normal,
+        Rotate90DegreesCounterclockwise,
+        Rotate270DegreesCounterclockwise
+    }
+
+    /// <summary>
+    /// Provides event data for RefreshRequested events.
+    /// </summary>
+    public class RefreshRequestedEventArgs : RoutedEventArgs
+    {
+        private RefreshCompletionDeferral _refreshCompletionDeferral;
+
+        /// <summary>
+        /// Gets a deferral object for managing the work done in the RefreshRequested event handler.
+        /// </summary>
+        /// <returns>A <see cref="RefreshCompletionDeferral"/> object</returns>
+        public RefreshCompletionDeferral GetDeferral()
+        {
+            return _refreshCompletionDeferral.Get();
+        }
+
+        public RefreshRequestedEventArgs(Action deferredAction, RoutedEvent? routedEvent) : base(routedEvent)
+        {
+            _refreshCompletionDeferral = new RefreshCompletionDeferral(deferredAction);
+        }
+
+        public RefreshRequestedEventArgs(RefreshCompletionDeferral completionDeferral, RoutedEvent? routedEvent) : base(routedEvent)
+        {
+            _refreshCompletionDeferral = completionDeferral;
+        }
+
+        internal void IncrementCount()
+        {
+            _refreshCompletionDeferral?.Get();
+        }
+
+        internal void DecrementCount()
+        {
+            _refreshCompletionDeferral?.Complete();
+        }
+    }
+
+    /// <summary>
+    /// Deferral class for notify that a work done in RefreshRequested event is done.
+    /// </summary>
+    public class RefreshCompletionDeferral
```

Comment:
Since you have moved these controls into the separated folder, you can move these types out to their own files.

Reply:
Including enums